cmd/atlas/internal/cmdapi/migrate.go

```diff
@@ -350,44 +354,128 @@ func CmdMigrateApplyRun(cmd *cobra.Command, args []string) error {
 		return err
 	}
 	var (
+		mux = tx{ctx: cmd.Context(), c: c, rrw: rrw, files: pending}
 		drv migrate.Driver
-		tx  *sqlclient.TxClient
 	)
-	if MigrateFlags.Apply.DryRun {
-		drv = &dryRunDriver{c.Driver}
-		rrw = &dryRunRevisions{rrw}
-	}
 	for _, f := range pending {
-		if !MigrateFlags.Apply.DryRun {
-			// Wrap the file execution in a transaction.
-			tx, err = c.Tx(cmd.Context(), nil)
-			if err != nil {
-				return err
-			}
-			drv = tx.Driver
+		drv, rrw, err = mux.driver(cmd.Context())
+		if err != nil {
+			return err
 		}
 		ex, err := migrate.NewExecutor(drv, dir, rrw, executorOptions(l)...)
 		if err != nil {
 			return err
 		}
-		if err := ex.Execute(cmd.Context(), f); err != nil {
-			if !MigrateFlags.Apply.DryRun {
-				if err2 := tx.Rollback(); err2 != nil {
-					err = fmt.Errorf("%v: %w", err2, err)
+		if err := mux.mayRollback(ex.Execute(cmd.Context(), f), f); err != nil {
+			return err
+		}
+		if err := mux.mayCommit(); err != nil {
+			return err
+		}
+	}
+	if err := mux.commit(); err != nil {
```

Comment:
My concern about using a separate connection for the revision-writer is that we lose the atomicity of the database. I think we can use SAVEPOINT to have both the information in the revisions table, but also maintain consistency. 

This of course can be added in future PRs because that’s the current behavior..